node/src/components/linear_chain_sync.rs

```diff
@@ -117,18 +247,62 @@ impl<I: Clone + 'static> LinearChainSync<I> {
         REv: ReactorEventT<I>,
     {
         let peer = self.random_peer_unsafe(rng);
-        match self.linear_chain.pop() {
+
+        let curr_state = mem::replace(&mut self.state, State::None);
+
+        let (next_state, next_block) = match curr_state {
+            State::None | State::Done => {
+                panic!("Tried fetching next block when in {:?} state.", curr_state)
+            }
+            State::SyncingTrustedHash {
+                trusted_hash,
+                highest_block_seen,
+                mut linear_chain,
+                ..
+            } => {
+                let next_block = linear_chain.pop();
+                let new_state = State::SyncingTrustedHash {
+                    trusted_hash,
+                    highest_block_seen,
+                    linear_chain,
+                };
+                (new_state, next_block)
+            }
+            State::SyncingDescendants {
+                ref linear_chain,
+                trusted_hash,
+                highest_block_seen,
+            } => {
+                if linear_chain.is_empty() {
+                    (curr_state, None)
+                } else {
+                    // We want to download deploys from lowest to highest block height.
+                    // When downloading descendants of trusted hash,
+                    // last element in the `linear_chain` has the highest height so we cannot use
+                    // `pop()`.
+                    let (head, tail) = linear_chain.split_at(1);
```

Comment:
Ha. Ok :+1:

Reply:
Looks like it will be less performant and will actually need more code (b/c we have to handle different collection types differently in some `match` cases).

Are you OK with leaving it as is?

Reply:
Taking ownership of `linear_chain` above and calling `linear_chain.pop_front()` would certainly be more performant, wouldn't it? (There's no reason for the `ref`.)

But if we have to I'm fine with this as well, of course.

Reply:
I looked at the body of `pop_front()` (and corresponding `push_back()` and `push_front()` on `VecDeque`) and it looked like it has many more operations necessary to get the element from the front.

> There's no reason for the `ref`

I can try but I thought that clippy would have told me about it.

Reply:
Done.